### PR TITLE
ContractModel: wellformedness checks for Actions

### PR DESCRIPTION
**Pull request by MaximilianAlgehed**
This PR introduces some checks that actions are well formed and don't refer to symbolic tokens that don't exist. A `Data.Data` based automatic predicate checks that no symtokens that aren't bound are used in actions and there is a possibility for the programmer to override this functionality with their own custom function (in case there is an issue with `Data.Data`). This means that issues like the one encountered in #342 don't affect developers and my "workaround" there isn't necessary.

Furthermore, this PR updates the dependency on `plutus-core` because we need instances of `Data.Data` for a number of types in `plutus-core`.

Pre-submit checklist:
- Branch
    - [x] Tests are provided (if possible)
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Relevant tickets are mentioned in commit messages
    - [x] Formatting, materialized Nix files, PNG optimization, etc. are updated
- PR
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested

### CODE REVIEW

**Comment by MaximilianAlgehed** [comment]
@sjoerdvisscher when you have the time this is ready :)

**Review by sjoerdvisscher — APPROVED** [review]

File: plutus-contract/src/Plutus/Contract/Test/ContractModel/Internal.hs [hunk 1]
```diff
@@ -392,6 +398,12 @@ type SpecificationEmulatorTrace a =
 -- `ModelState` type, which in addition to @state@ tracks common features of the blockchain, like
 -- wallet balances and the current slot.
 
+class (Eq (Action state), Show (Action state)) => HasActions state where
+  getAllSymtokens :: Action state -> Set SymToken
+
+instance {-# OVERLAPPABLE #-} (Eq (Action state), Show (Action state), Data (Action state)) => HasActions state where
+  getAllSymtokens = Set.fromList . universeBi
```
[the remark this answers is not in the document]
Comment (MaximilianAlgehed) [reply plutus-contract/src/Plutus/Contract/Test/ContractModel/Internal.hs:405]: I looked into this (I didn't know about `DefaultSignatures`) and I don't see how they replace overlapping instances? As far as I can tell using this extension would require the programmer to state that they want an instance for `HasActions` for their datatype (possibly via `derive`?) whereas now you don't have to add this to your `deriving` list. Instead, if for some very specific reason you need to overwrite the default instance (because you can't rely on `Data` for whatever reason) you would now be required to add an `OVERLAPPING` pragma.

In other words: doing this would mean more work and more understanding in the common cases to have slightly less work (and very slightly less understanding) in the uncommon cases. I think I will pass on this suggestion here unless there is something I'm misunderstanding?

[the remark this answers is not in the document]
Comment (sjoerdvisscher) [reply plutus-contract/src/Plutus/Contract/Test/ContractModel/Internal.hs:405]: There must be a downside to doing it this way, since I haven't seen this pattern before, have you? 
Maybe this didn't work well with the old `OverlappingInstances`?
I don't think it is a performance issue.
Or maybe it's just that in other cases it is more common to write you own instance and then having to add `OVERLAPPING` could be annoying.

[the remark this answers is not in the document]
Comment (MaximilianAlgehed) [reply plutus-contract/src/Plutus/Contract/Test/ContractModel/Internal.hs:405]: I think the last thing you wrote is most likely - with the new overlapping rules I don't see much room for confusion.ö
